### ISSUE TITLE
"position" handler and no transitions issue

### ISSUE THREAD

**Issue by valichek** [aurelia/dialog#190]
When "position" handler is being set and no transitions for dialog defined in css, the promise returned from DialogRenderer.showDialog() resolves on dialog CLOSE! transition end.

**Comment by EisenbergEffect** [comment]
@jedd-ahyoung Ping

**Comment by EisenbergEffect** [comment]
@PWKad Are you able to look at this?

**Comment by don-bluelinegrid** [comment]
Curious what the status is of this or https://github.com/aurelia/dialog/issues/208

We need to be able to do the following:
1. Open a dialog.
2. After user clicks 'OK', execute a fetch-client REST request.
3. On success of the request based on the fetch-client resolved Promise, then close the dialog; on failure of the request, show an error message in the dialog, which *should still be open*.

My understanding is that openAndYieldController() is designed for this use case, I.e., #3.  I've used code from the example for this, but the dialog closes immediately on calling ok() from the OK button click.  Is what I've described the same case as what's reported here, or something different?

Is there some other way to accomplish this?

Thanks,
Don

**Comment by valichek** [comment]
@don-bluelinegrid to avoid this issue use `{ ignoreTransitions: true }` for options

**Comment by valichek** [comment]
@don-bluelinegrid actually that depends on what your OK btn is doing, so smth different

**Comment by don-bluelinegrid** [comment]
@valichek 

1. Where do we set ```{ ignoreTransitions: true }```?

2. 

> actually that depends on what your OK btn is doing, so smth different
My OK button is just calling ok().  How else can I use the OK button to communicate with teh DialogController to produce a result inside the Promise?

**Comment by don-bluelinegrid** [comment]
@valichek 

I've tried ignoreTransitions, it made no difference with the code I am using.

Here is the code in my app:

```
    this.dialogService.openAndYieldController({
      viewModel: Model, 
      view: './communityModel.html', 
      model: {title: title, item: community, okText: this.i18n.tr('button.save')},
      ignoreTransitions: true 
    }).then((controller:any) => {
       controller.result.then((response) => {
    // .then(response => {
      if (!response.wasCancelled) {
        let community = response.output;
        if(community.commnunityId ) {
          // Update
        } else {
          // Create
          let comm = {
            communityId: community.communityId, 
            communityName: community.communityName, 
            communityDescription: community.communityDescription, 
            communityType: community.communityType,
            membershipType: 'DEFINED'
          };
          this.dataService.createCommunity(comm)
            .then(response => response.json())
            .then(data => {
              let res = data;
            }, error => {
              console.debug("Community create() rejected."); 
            }).catch(error => {
              console.debug("Community create() failed."); 
              console.debug(error); 
              return Promise.reject(error);
            })
        }
        // Call the delete service.
        console.debug('Edited');
        console.debug(response.output);
      } else {
        // Cancel.
        console.debug('Cancel');
      }
    })
    });


  }

```

In my HTML template, I am simply calling ok(output).  This immediately closes the dialog.

**Comment by valichek** [comment]
@don-bluelinegrid  options for `openAndYieldController ()`
It closes dialog, yes, and then on ok received you're sending the request, to send request first before closing you need to set smth like 
```
let model = controller.settings.model;
model.submit = () => {
    this.dataService.createCommunity().then(() => {
        controller.ok()
    })
}
controller.result.then()...
```

and call `model.submit()` in `communityModel.html`

**Comment by don-bluelinegrid** [comment]
@valichek 

Thanks for the information.

One clarification -
It sounds like you are saying that for this case, using openAndYieldController() is not necessary?  So I should go back to the normal open() call?

Or do I need to use openAndYieldController() so that I can add the submit() function to the controller as the dialog opens?

Thanks,
Don

**Comment by valichek** [comment]
@don-bluelinegrid if you go back to `open()` you need to call `createCommunity ()` from Model, because it resolves on dialog result, `openAndYieldController()` resolves on actual dialog opening

**Comment by don-bluelinegrid** [comment]
@valichek 

Thanks very much for the information/advice.  Your approach does work, and I have things working as intended now, to wait for a successfully resolved Promise from the fetch call before closing the dialog.

One comment -
The reason for doing this is to have the ability to display an error/alert message inside the dialog, in the case of a failure.  This is a very common UX pattern.  It would be nice to have some built-in support for this, because with the current approach there is a fair amount of boilerplate that must be added to the dialog-opening function; specifically, the code that adds the submit() function and processes that response.

An enhancement would be to have an additional built-in controller function, like okNoClose(), which submits the output without closing the dialog.

Thanks,
Don

**Comment by huochunpeng** [comment]
Even at end of dialog close, the promise is not always resolved #229.

Suggest to make the default value of `ignoreTransitions` as `true` for a temporary fix, so others (like me) would not spend a day to tracing through code to understand what's going on.
